skimage/feature/_haar.pyx

```diff
@@ -0,0 +1,318 @@
+#cython: cdivision=True
+#cython: boundscheck=False
+#cython: nonecheck=False
+#cython: wraparound=False
+
+from libc.stdlib cimport malloc, free, realloc
+
+import numpy as np
+
+from ..transform import integral_image
+from .._shared.transform cimport integrate
+
+FEATURE_TYPE = {'type-2-x': 0, 'type-2-y': 1,
+                'type-3-x': 2, 'type-3-y': 3,
+                'type-4': 4}
+
+N_RECTANGLE = {'type-2-x': 2, 'type-2-y': 2,
+               'type-3-x': 3, 'type-3-y': 3,
+               'type-4': 4}
+
+
+cdef Rectangle** _haar_like_feature_coord(Py_ssize_t width,
+                                          Py_ssize_t height,
+                                          unsigned int feature_type,
+                                          Py_ssize_t* n_rectangle,
+                                          Py_ssize_t* n_feature) nogil:
+    """Private function to compute the coordinates of all Haar-like features.
+    """
+    cdef:
+        # allocate for the worst case scenario
+        Py_ssize_t max_feature = height ** 2 * width ** 2
+        Rectangle** rect_feat = NULL
+        Py_ssize_t cnt_feat = 0
+        Py_ssize_t local_n_rectangle
+        Py_ssize_t x, y, dx, dy
+
+    if feature_type == 0 or feature_type == 1:
+        local_n_rectangle = 2
+    elif feature_type == 2 or feature_type == 3:
+        local_n_rectangle = 3
+    else:
+        local_n_rectangle = 4
+    n_rectangle[0] = local_n_rectangle
+
+    rect_feat = <Rectangle**> malloc(local_n_rectangle * sizeof(Rectangle*))
```

Comment:
How much slower is it to use a Cython list, so that we don't have to do any custom memory allocation?

Reply:
I'll try to check. Not sure what you mean by cython list. If actually you require a python interaction then I will not be able to release the GIL, which could be interesting in case you want to make Parallel with threading backend.

Reply:
So you go from 150 ms to 240 ms to compute 43,000 features.

I put the code in a [gist](https://gist.github.com/glemaitre/c5f31b0669c1c11c8e2de272dcb4a665#file-list_haar-py). 
Appending in a python list will trigger a conversion of the C structure to a python dictionary.

Personally, I prefer a pure C allocation because I know how C will behave. To be honest, I was unsure of what would be the implicit conversion and I am not sure that mixing python and C in this specific piece of code will make it easier to understand while reading. But I am not an expert in Cython, so it could be due to make lack of experience there.

I still think that releasing the GIL could be nice as well.

WDYT?

Reply:
@glemaitre You can also try C++ vector like here:
https://github.com/warmspringwinds/scikit-image/blob/object_detect_module/skimage/future/detect/cascade.pyx#L92